### ISSUE TITLE
Deadlocked "MQTT Rec:" When "MQTT Call:" Dies

### ISSUE THREAD

**Issue by jpwsutton** [eclipse-paho/paho.mqtt.java#47]
migrated from Bugzilla [#443142](https://bugs.eclipse.org/bugs/show_bug.cgi?id=443142)
status RESOLVED severity _blocker_ in component _MQTT-Java_ for _1.1_
Reported in version _unspecified_ on platform _All_
Assigned to: Bin Zhang

Original attachment names and IDs:
- _[443142 Test-case.java.txt](https://bugs.eclipse.org/bugs//attachment.cgi?id=247164)_ (ID 247164)

On 2014-09-02 15:56:42 -0400, Gary Russell wrote:

> See my analysis here: http://stackoverflow.com/questions/25620196/spring-integration-mqtt-subscriber-paho-stops-processing-messages/25631077# 25631077
> 
> When an exception is thrown on the "Call" thread, and the "Rec" thread is blocked because the queue is full, the connection is shutdown but that causes a deadlock because the "Snd" thread waits for the "Rec" thread, which will never exit.
> 
> It looks like the 'spaceAvailable.notifyAll()' needs to go into a finally block.

On 2014-09-02 15:57:27 -0400, Gary Russell wrote:

> This is with the 0.4.0 client, but it looks the same on master.

On 2014-09-02 17:00:13 -0400, Gary Russell wrote:

> I see the same problem in 'CommSender.handleRunException()`. If the sender throws an exception, it calls`clientComms.shutDownConnection()` which, again waits for the 'Rec:' thread to stop, which will never happen because he is blocked on 'spaceAvailable'.

On 2014-09-15 04:06:59 -0400, Bin Zhang wrote:

> Could you please paste the stack trace when using the paho java client code on master?

On 2014-09-17 04:36:11 -0400, Gary Russell wrote:

> I won't be able to get to this until at least Sept 25th.
> 
> I will see if I can have a colleague help you.
> 
> However, it was easy to reproduce a similar deadlock (in the 0.4 released client - I didn't try on master).
> 
> Send > 10 messages
> 
> In MessageArrived() use
> 
> Thread.sleep(1000); // so the Rec: queue is full and the thread suspended
> throw SomeException()
> 
> The Call: thread is deadlocked waiting to join the Rec: thread.

On 2014-09-17 06:13:59 -0400, Bin Zhang wrote:

> Unfortunately, i failed to reproduce the issue according the steps you described if i understand correctly with code on masterã
> 
> It would be great if your colleague can provide a test to reproduce this issue with code on master, since several deadlock bugs were fixed since 0.4 release.
> 
> Thanks.

On 2014-09-17 13:03:24 -0400, Artem Bilan wrote:

> Created attachment 247164
> Test-case to reproduce
> 
> See test-case in the attachment. Reproducable on the 0.4.0 as well as on the 0.4.1-SNAPSHOT.
> As Gary said there is need to change this code in the CommsCallback:
>   synchronized (spaceAvailable) {
>                   // Notify the spaceAvailable lock, to say that there's now
>                   // some space on the queue...
> 
> ```
>               // @TRACE 706=notify spaceAvailable
>               log.fine(CLASS_NAME, methodName, "706");
>               spaceAvailable.notifyAll();
>           }
>       } catch (Throwable ex) {
>           // Users code could throw an Error or Exception e.g. in the case
>           // of class NoClassDefFoundError
>           // @TRACE 714=callback threw exception
>           log.fine(CLASS_NAME, methodName, "714", null, ex);
>           running = false;
>           clientComms.shutdownConnection(null, new MqttException(ex));
>       }
> ```
> 
> when we have to move spaceAvailable.notifyAll(); to finally block.

On 2014-09-18 12:06:11 -0400, Bin Zhang wrote:

> Tried moving spaceAvailable.notifyAll(); to finally block, the test still failed.
> 
> CommsCallback calls ClientComms.shutdownConnection() once it caught some exception in the callback thread,
> and it stops networkModule which closes the socket, and then CommsReceiver will got SocketException and running=false, thread stopped.
> 
> Regarding this condition:  while (!quiescing && messageQueue.size() >= INBOUND_QUEUE_SIZE) { spaceAvailable.wait(); ... , 
> even spaceAvailable.notifyAll() is called, the while condition is still true, it continues waiting.
> 
> Actually, i think it's not reasonable an exception thrown from the MqttCallback that causes the MQTT client to shutdown.
> 
> Any ideas?

On 2014-09-19 08:48:53 -0400, Artem Bilan wrote:

> Eh... Sorry, now I see that my test-case doesn't show the real issue.
> It wasn't really clear from Gary's comments.
> 
> However I agree that it is very aggressive to stop Client on an Exception from the target application's MqttCallback implementation.

On 2014-09-29 05:18:31 -0400, Gary Russell wrote:

> This reproduces a similar deadlock:
> 
> https://gist.github.com/garyrussell/80577e9e8eb865aa37f0
> 
> Deadlocks with the released (0.4.0) client as well as current GitHub master.
> 
> After running the program...
> 
> $ ps -ef | grep Deadlock | grep -v grep | awk '{ print $2 }' | xargs jstack
> 
> ...
> 
> "MQTT Call: bar" prio=5 tid=0x00007fedf1879000 nid=0x4c03 in Object.wait() [0x000000011fe2e000]
>    java.lang.Thread.State: WAITING (on object monitor)
>   at java.lang.Object.wait(Native Method)
> - waiting on <0x000000010a1245b8> (a java.lang.Thread)
>   at java.lang.Thread.join(Thread.java:1260)
> - locked <0x000000010a1245b8> (a java.lang.Thread)
>   at java.lang.Thread.join(Thread.java:1334)
>   at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.stop(CommsReceiver.java:83)
> - locked <0x000000010a1243a8> (a java.lang.Object)
>   at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:307)
>   at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:184)
>   at java.lang.Thread.run(Thread.java:724)
> 
> "MQTT Rec: bar" prio=5 tid=0x00007fedf288f000 nid=0x4a03 in Object.wait() [0x000000011fc28000]
>    java.lang.Thread.State: WAITING (on object monitor)
>   at java.lang.Object.wait(Native Method)
> - waiting on <0x000000010a111350> (a java.lang.Object)
>   at java.lang.Object.wait(Object.java:503)
>   at org.eclipse.paho.client.mqttv3.internal.CommsCallback.messageArrived(CommsCallback.java:299)
> - locked <0x000000010a111350> (a java.lang.Object)
>   at org.eclipse.paho.client.mqttv3.internal.ClientState.notifyReceivedMsg(ClientState.java:864)
>   at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:127)
>   at java.lang.Thread.run(Thread.java:724)
>   ...
> 
> Note that the Call: thread is waiting to join the Rec: thread, which is waiting on spaceAvailable.

On 2014-09-30 06:37:49 -0400, Bin Zhang wrote:

> The root cause is the CommsCallback.messageArrived, the while
> condition will never be false when the messageQueue is full and not quiescing, even when the callback thread is trying to stop.
> 
> I will fix this.

On 2014-09-30 06:50:43 -0400, Bin Zhang wrote:

> Fixed in develop branch via https://git.eclipse.org/r/#/c/34109/

On 2014-10-28 12:22:45 -0400, Gary Russell wrote:

> Any plans to release this fix? I only see the old 0.4 release in maven: https://repo.eclipse.org/content/repositories/paho-releases/org/eclipse/paho/mqtt-client/

On 2015-05-06 04:17:50 -0400, Anand  J. Kadhi wrote:

> So if this issue is fixed in which release can we get the fixed version, actually currently we are using spring-integration-mqtt 4.1.3 version which internally uses the mqtt-client 0.4.0, and we are facing the same problem. can you please tell where can i get the jar with the above fix?

On 2015-05-06 04:46:28 -0400, Gary Russell wrote:

> The maven artifactId has changed. I just tested Spring Integration against the 1.0.2 release.
> 
> Repo: https://repo.eclipse.org/content/repositories/paho-releases/
> GroupId: org.eclipse.paho
> ArtifactId: org.eclipse.paho.client.mqttv3

On 2015-05-06 06:01:27 -0400, Davy De Waele wrote:

> Gary,
> 
> If I understand correctly the latest version of Spring Integration still uses the (very) old 0.4.0 paho mqtt client.
> 
> Are there plans to align Spring integration with a more recent version of Paho.
> 
> I have the impression lots of people are confused / unaware of the maven groupId/artifactId change and most of them are still on 0.4.0
> 
> Would it be a good idea to exclude the 0.4.0 and replace it with a more recent 1.0.x release and test it against spring integration 4.1.3 ?

On 2015-05-06 06:12:02 -0400, Gary Russell wrote:

> Yes, we have an open JIRA issue [1] to upgrade, but we're having some problems pulling it into the spring.io repository; it's some problem with the parent pom group id and the way our cache is set up. We like to do that so users don't have to worry about adding the eclipse repo to their project.
> 
> I ran the test suite this morning with the 1.0.2 artifact. So, yes, until INT-3645 is resolved, I suggest you do as you say (exclude the old artifact and pull in the new one from the eclipse repo).
> 
> Please report any issues you find (that are Spring Integration related) using our JIRA.
> 
> Thanks
> 
> [1]: https://jira.spring.io/browse/INT-3645

On 2015-05-06 06:14:51 -0400, Ian Craggs wrote:

> We tried removing 0.4 from the Nexus repository last year, to stop new implementations using it, and to make people aware of the change in name.  That caused an outcry so we had to put it back.
> 
> Unfortunately, the package name had to change when the first official Eclipse release was made, to conform to Eclipse naming conventions.

On 2015-05-06 06:22:41 -0400, Davy De Waele wrote:

> Hi Ian,
> 
> I understand, once released it's not trivial to remove.
> 
> But I would like to urge the Eclipse Paho site maintainers to highlight their note on the groupId/artidactId in a larger font and in red :)

On 2015-05-06 06:29:09 -0400, Ian Craggs wrote:

> (In reply to Davy De Waele from comment # 18)
> 
> > Hi Ian,
> > 
> > I understand, once released it's not trivial to remove.
> > 
> > But I would like to urge the Eclipse Paho site maintainers to highlight
> > their note on the groupId/artidactId in a larger font and in red :)
> 
> Should I make it blink as well?

On 2015-05-06 06:52:58 -0400, Ian Craggs wrote:

> I've moved the message, to make it more prominent, added version numbers, and turned it red.  But I haven't made it blink :-)

On 2015-05-06 07:44:40 -0400, Davy De Waele wrote:

> No worries, I'll log another issue to make it blink :)

On 2015-05-07 01:31:32 -0400, Anand  J. Kadhi wrote:

> Hi team,
> Thanks for the help i have replaced the maven groupId and artifactId now it's working fine. Great.
